utils/python/hsfs_utils.py

```diff
@@ -325,9 +326,21 @@ def offline_fg_materialization(spark: SparkSession, job_conf: Dict[Any, Any], in
     # deserialize dataframe so that it can be properly saved
     deserialized_df = engine.get_instance()._deserialize_from_avro(entity, filtered_df)
 
+    # de-duplicate records
+    # timestamp cannot be relied on to order the records in case of duplicates, if they are produced together they would have the same timestamp.
+    # Instead use offset to order the records, they are strictly increasing within a partition and since we use primary keys for generating Kafka message keys duplicates are guaranteed to be in the same partition.
+    window = Window.partitionBy([f"value.{key}" for key in entity.primary_key]) \
+                .orderBy(col("offset").desc())
+    deduped_df = deserialized_df.withColumn("row_num", row_number().over(window)) \
+                .filter("row_num = 1") \
+                .drop("row_num")
+
+    # get only the feture values (remove kafka metadata)
+    deduped_df = deduped_df.select("value.*")
+
     # insert data
     entity.stream = False # to make sure we dont write to kafka
-    entity.insert(deserialized_df, storage="offline")
+    entity.insert(deduped_df, storage="offline")
 
     # update offsets
     df_offsets = (df if limit > filtered_df.count() else filtered_df).groupBy('partition').agg(max('offset').alias('offset')).collect()
```

Comment:
Not specifically related to this PR - but this `filtered_df` is not cached - so the `count()` might recompute the thing from the beginning and might be different from what it was written in the offline store.

Are we missing some PRs here?